src/cli/commands/force/lightning/local/device/create.ts

```diff
@@ -0,0 +1,258 @@
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { flags, FlagsConfig } from '@salesforce/command';
+import { Messages, SfdxError } from '@salesforce/core';
+import cli from 'cli-ux';
+import util from 'util';
+import { AndroidSDKUtils } from '../../../../../../common/AndroidUtils';
+import { CommandLineUtils } from '../../../../../../common/Common';
+import { CommonUtils } from '../../../../../../common/CommonUtils';
+import { IOSUtils } from '../../../../../../common/IOSUtils';
+import { Requirement } from '../../../../../../common/Requirements';
+import androidConfig from '../../../../../../config/androidconfig.json';
+import Setup from '../setup';
+
+// Initialize Messages with the current plugin directory
+Messages.importMessagesDirectory(__dirname);
+
+// Load the specific messages for this file. Messages from @salesforce/command, @salesforce/core,
+// or any library that is using the messages framework can also be loaded this way.
+const messages = Messages.loadMessages(
+    '@salesforce/lwc-dev-mobile',
+    'devicecreate'
+);
+
+// android: createNewVirtualDevice(emulatorName: emu21, emulatorImage: google-apis, platformAPI: android-29, device: pixel, abi: x86_64)
+// ios: createNewDevice(simulatorName: mydevice, deviceType: iPhone-8, runtime: iOS-14-2)
+
+export default class Create extends Setup {
+    public static description = messages.getMessage('commandDescription');
+
+    public static readonly flagsConfig: FlagsConfig = {
+        devicename: flags.string({
+            char: 'n',
+            description: messages.getMessage('deviceNameFlagDescription'),
+            required: true
+        }),
+        devicetype: flags.string({
+            char: 'd',
+            description: messages.getMessage('deviceTypeFlagDescription'),
+            required: true
+        }),
+        platform: flags.string({
+            char: 'p',
+            description: messages.getMessage('platformFlagDescription'),
+            required: true
+        })
+    };
+
+    public static instance: Create;
```

Comment:
I had to add this b/c device create has 2 requirements in addition to the base setup requirements. The functions for verifying these 2 requirements need to access property values of this class, but when we are inside those functions, we don't have access to `this.<property>` because in the context of those functions `this` refers to the `Requirement` class and not `Create` class.

Reply:
Yeah, this has unfortunately been a recurring side effect of how we've defined our `Requirement` instances to date, in the "literal class" inline array definitions. It gets messy enough that I wonder if it's something we can refactor before GA.

If we defined our `Requirement` implementors in terms of concrete classes that implement the interface, it would free us up to encapsulate the "owner" object as a member of the concrete class.  Consider something like:

```typescript
type CheckRequirementsFunc = () => Promise<string>;

interface Requirement {
    title: string;
    checkFunction: CheckRequirementsFunc;
    resolutionMessage: string;
}

class RequirementOwner {
    public ownerProperty: string;
}

class SpecificRequirement implements Requirement {
    public title: string = "MySpecificRequirement";
    public resolutionMessage: string = "Resolved! Owner property is";
    private owner: RequirementOwner;

    constructor(owner: RequirementOwner) {
        this.owner = owner;
    }

    public checkFunction(): Promise<string> {
        return new Promise<string>((resolve, reject) => {
            resolve(`${this.resolutionMessage} ${this.owner.ownerProperty}.`);
        });
    }
}

// ...

const requirements: Requirement[] = [
    new SpecificRequirement(owner)
];
```

It's not far off of the ways we've implemented this so far in the project, but it's a lot cleaner in its intent and encapsulation. Not necessarily a task for this PR, but I'd like to know what you and @sfdctaka think about the adjustment.

Reply:
Actually, we don't even have to change it everywhere to start. I'd propose we make a go of it starting with this class specifically, since it's new. So e.g.:

```typescript
const extraReqs: Requirement[] = [
    {
        checkFunction: this.checkUniqueDevice,
        fulfilledMessage: messages.getMessage(
            'reqs:devicename:fulfilledMessage'
        ),
        logger,
        title: messages.getMessage('reqs:devicename:title'),
        unfulfilledMessage: messages.getMessage(
            'reqs:devicename:unfulfilledMessage'
        )
    },
// ...
];
```

Becomes:

```typescript
const extraReqs: Requirement[] = [
    new UniqueDeviceRequirement(this), // owner arg, checkUniqueDevice() gets defined as checkFunction() in UniqueDeviceRequirement, etc.
//... etc.
];
```

These requirements classes can be inline-defined within `create.ts`. Thinking it over, I'd rather take that approach than the workarounds that we're currently implementing, which feel like they're reaching to make up for a deficiency in our code structure. Thoughts?

Reply:
I like Kevin's second approach. Maybe there's a third approach that we can try. I need to try but something like
```
const extraReqs: Requirement[] = [
    {
         checkFunction: this.checkUniqueDevice.bind(this),
```
where `bind` will explicitly make this in `checkUniqueDevce` an instance of `Create`. I'll give it a try and let you know if it works.

Reply:
A quick test using `bind` worked (and I removed all lines with `Create.instance`).
![image](https://user-images.githubusercontent.com/42476999/104532183-bf2f7f00-55b3-11eb-9c5f-d90ac7f6d6be.png)

![image](https://user-images.githubusercontent.com/42476999/104532200-c8205080-55b3-11eb-82ec-bb3a19747967.png)

`bind` is a very language specific feature and not often used. If we have work around like in Kevin's 2nd approach or refactoring(before GA) I am more inclined to give them a higher consideration.

Reply:
Kevin, I had implemented something similar to this at first but decided to revert it b/c I thought maybe it is not a good approach 🤦‍♂️ (thinking that if it is not implemented carefully it might lead to cyclic dependency and memory issues etc.)

Basically what I had done was to update the `Requirement` definition to this:

```typescript
export interface Requirement {
    title: string;
    checkFunction: CheckRequirementsFunc;
    fulfilledMessage: string;
    unfulfilledMessage: string;
    logger: Logger;
    owner: any;
}
```

But then in my approach `owner` was of type `any` and the consumer would have to cast it again, whereas in your approach that is not needed..... so I like your approach.

Takashi, thanks for verifying. That's pretty neat too.